src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -897,8 +897,20 @@ public boolean isComplete() {
         // TODO: if program terminates with exception, we need to record it
         // TODO: in the error case, we have to close all the open nodes
         FlowNode head = new FlowEndNode(this, iotaStr(), (FlowStartNode)startNodes.pop(), result, getCurrentHeads().toArray(new FlowNode[0]));
-        if (outcome.isFailure())
-            head.addAction(new ErrorAction(outcome.getAbnormal()));
+        if (outcome.isFailure()) {
+            try {
+                head.addAction(new ErrorAction(outcome.getAbnormal()));
+            } catch (Exception e) {
```

Comment:
> I am not sure we can do much to recover in those cases

Then again, the originally reported problem was less about the assertion failing to be printed, than the whole build hanging, so something does need to be fixed here just in case something else does go terribly wrong.

But I suspect `CpsThreadGroup` is not the best place to do this. Rather, [here](https://github.com/jenkinsci/workflow-api-plugin/blob/8ca82ae8180dc0869f083f2684c291e2faa18222/src/main/java/org/jenkinsci/plugins/workflow/graph/FlowNode.java#L301-L305) we are catching `IOException` but not the `RuntimeException`s that are common with XStream. So this whole PR could probably be closed and replaced with one to `workflow-api` which both patches `ErrorAction` to properly save blacklisted power assertions (with full details), *and* handles any other unforeseen problems adding `ErrorAction` or anything else to a node, and also includes a functional test reproducing the bug.